### DISCUSSION TITLE
Need Help on creating a popup to capture user input only while Uninstallation

### DISCUSSION THREAD

**Discussion by jchirantan** [wixtoolset/issues#8015] [Questions] [answered]
Project - Java based
Tool - Wix
Please let me know if further information is required.
Any sample code snippet will be helpful.

Thanks in Advance,
Chirantan

**Comment by jchirantan** [comment]
Any update on the above query?

Thanks,
Chirantan

**Comment by jchirantan** [comment]
Any update on the above query?

Thanks,
Chirantan

**Comment by chrpai** [comment]
Microsoft doesn't recommend any UI during the uninstall. In fact  uninstall has no UI.  You'd have to write a custom action.  What is the goal?  I've seen scenarios where people want to enforce a password protection of uninstall (block if invalid) ,  I've seen people want to prompt for save user data (Microsoft answer is always save)   and I've seen people just want to get feedback on why are you leaving us.

**Comment by jchirantan** [comment]
Thanks for your reply.
I want to prompt user for full uninstall or partial uninstall.
Partial Uninstall: User wants to keep cache and files created while using product. 
Full Uninstall: User wants to uninstall product by removing all the related files and cache.

Can we do this?

Thanks in advance,
Chirantan

**Comment by chrpai** [comment]
Is your installer just an MSI using Windows Installer Native UI or are you using a bootstrapper (.exe) as a wrapper?

**Comment by jchirantan** [comment]
I am working on the enhancement of existing project and have very little information about the wix implementation in the project.

I think it must be a bootstrapper, as the project has an XML configuration file named MFPBootstrapper.wixproj
Also, I have Bundle.wxs and Product.wxs files which are referenced from ".wixproj" file.

**Comment by chrpai** [comment]
Yes that's a bootstrapper. In that case the bootstrapper application handles the command line params, UI and calling of the MSI.  So you would extend it to have the functionality that you want.

**Comment by jchirantan** [comment]
How can I do that?

**Reply by chrpai** [reply]
I have even less information about the wix implementation in the project then you do as I've never seen it.  Either learn WiX or hire a consultant.  (Like myself.)

**Comment by jchirantan** [comment]
Thank you for your valuable insights.
Also, is there any video or documentation reference where I can see and understand about wix and extending wix for custom changes?

**Comment by chrpai** [comment]
This might be helpful to you.   The first 30 minutes in an overview of WiX and the last 20 minutes gets into bootstrappers specifically.

https://www.youtube.com/watch?v=XfwlWUJPmYU

WiX v3 has a custom bootstrapper that you could look at the source.  There's also some OSS projects that create bootstrappers such as WiXWPF that you could look at.

Also a quick google search for "wix custom bootstapper"  gets you here: https://wixtoolset.org/docs/v3/bundle/ba/

**Comment by jchirantan** [comment]
Thank you so much for your help.

Thanks,
Chirantan

**Comment by jchirantan** [comment]
Hi @chrpai,

- I have created a dialog to accept the user input on uninstallation inside "Product.wxs" file.
![image](https://github.com/wixtoolset/issues/assets/73328079/6d7e406e-9524-427e-9dc9-31d7bd5c1b4d)

- Also, I have copy pasted code from WixUI_InstallDir.wxs file present on Wix git repo to "Product.wxs" and added my dialog in that code (I was just trying to add the dialog box at the installation).
![image](https://github.com/wixtoolset/issues/assets/73328079/e199b12e-835a-4979-8558-127727f8b6ca)

- Still I am not able to see the dialog box. 

Any help on above will be highly appreciated.

Thanks in advance,
Chirantan

**Comment by chrpai** [comment]
I'd need to see a minimal reproducible sample to tell you what your doing wrong.  Like a github repos I could fork and submit a pull request.

You said previously that you were using a bootstrapper but your code above is for MSI UI.  Bootstrappers don't generally display MSI UI.  Also MSI doesn't display any UI other then a progress bar when doing an uninstall.

**Comment by jchirantan** [comment]
Hi @chrpai,

Here is the git repo link for sample code.
Uninstallation and installation might not work properly as it is a partial installer code.

https://github.com/jchirantan/WiXProject

It will be highly appreciated, if you can visit the repo and help me with the pointers. 

Thanks, 
Chirantan

**Comment by chrpai** [comment]
I forked the repos but it doesn't build. It has dozens and dozens of missing files.  Strip that all out to make it the most simple install that demonstrates the problem.  Something that I can build, modify and submit a PR.

**Comment by jchirantan** [comment]
Thanks @chrpai for such a quick response.

By looking at the code structure and file names, can you tell if it is a bootstrapper project or a MSIUI project (i.e. can we extend this project to add UI at uninstallation)?

Thanks,
Chirantan

**Reply by wmanning** [reply]
It's a bundle.  Implied by the MFPBootstrapper.wixproj name and when I look in that file I see <OutputType>Bundle</OutputType>.

Bundle.wxs would be the file that contains all the wix bundle stuff.
Looks like it has a couple of ExePackage and a msi package in the chain. Those are the packages the bundle will install.

**Reply by jchirantan** [reply]
So, can we add UI dialog during uninstallation into this project?

**Reply by wmanning** [reply, marked as answer]
Custom UI is a much more advanced topic.  You are in for lots of fun.

We use wisstdba and have a custom theme but we don't have any custom controls so can't help much. 

You can use the standard bootstrapper application (wixStdBA) and add a custom theme file.  I was curious and did some research and it appears you can create a checkbox in the options page which can be linked to a burn variable and passed into the MSI.  Below links are older so no idea if anything has changed since then.  
https://stackoverflow.com/questions/11640470/burn-how-to-store-a-checkbox-value-in-a-variable
https://stackoverflow.com/questions/50570254/variable-value-isnt-changing-according-to-checkbox-value-from-rtf-theme 
https://github.com/wixtoolset/wix3/pull/162
https://github.com/wixtoolset/issues/issues/4568/

Second option is custom bootstrapper application (BA).  That might be even harder.  Significant changes from Wix v3 to v4 to v5. 

Seems like you need to get familiar with WiX as well?
Ten years ago this book is what I used to get up to speed on Wix: Wix 3.6: A Developer's Guide to Windows Installer XML, An In-and-Out, To-the-Point Introduction to Windows Installer XML.  It's quite old but still good for Wix v3.

Wix 4 came out early last year and it has significant changes.
https://wixtoolset.org/docs/fourthree/
https://wixtoolset.org/docs/fourthree/faqs/

Then now there's Wix 5 about to come out.

So a lot of learning to do.  Hope this is of some help to you or at least others.

**Reply by jchirantan** [reply]
Thanks @wmanning for your help.
I was able to add a checkbox in my uninstaller window by modifying theme.xml.

Also, Thanks @chrpai for your time and valuable insights.
Thanks again,
Chirantan

**Comment by robmen** [comment]
We're keeping questions open since they are easier to find in search.
